### PR TITLE
[jenkins] feat: setup private repository

### PR DESCRIPTION
**Pull request by Wayonb**
## What is the current behavior?
All packages are updated from the public artifact repository.
Projects need to push test artifacts to a public repository.

## What's the issue?
Currently have no support for private artifact repos.

## How have you changed the behavior?
The following have been added to Jenkins.
1. Add support to publish to private NPM, PyPi, and Docker repositories.
2. Add support to install packages from a private repository during CI
3. Private GitHub repository always publishes to private artifact repo.

The Nexus OSS version has certain limitations.  One been is that artifacts cannot be pushed to a group repository.
Jenkins will configure the pull from the group repo and push it to the hosted repo.

## How was this change tested?
Tested in Jenkins.

### CODE REVIEW

File: jenkins/shared-library/vars/publish.groovy [hunk 1]
```diff
@@ -55,49 +64,74 @@ void npmPublisher(Map config, String phase) {
...
-	writeFile(file: '.npmrc', text: '//registry.npmjs.org/:_authToken=${NPM_TOKEN}')
-	runScript('cat .npmrc')
-	withCredentials([string(credentialsId: NPM_CREDENTIALS_ID, variable: 'NPM_TOKEN')]) {
+	final String ownerName = helper.resolveOrganizationName()
+	final String repositoryName = helper.resolveRepositoryName()
+	if (isAlphaRelease(phase) || !isGitHubRepositoryPublic(ownerName, repositoryName)) {
+		final String publishUrl = configureArtifactRepository.resolveRepositoryUrl(ownerName, 'npm-hosted')
+		final String environment = jobHelper.resolveCiEnvironmentName(config)
+
+		npmPublishCommand += ' --registry=' + publishUrl
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/publish.groovy:73]: can use format string on lhs

File: jenkins/shared-library/vars/configureArtifactRepository.groovy [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	return userNamePassword
+}
+
+void configure(String environment, String gitOrgName, String url) {
+	Map artifactRepositoryInfo = [:]
+	Map<String, List<Object>> repositoryHandlerMap = [
+			'javascript' : this.&configureNpm,
+			'python': this.&configurePypi
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/configureArtifactRepository.groovy:96]: did you intend to double indent ?

File: jenkins/shared-library/vars/configureArtifactRepository.groovy [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		'javascript' : 'npm-group',
+		'python': 'pypi-group'
+	]
+
+	return environmentRepoNameMap[environment] ?: ''
+}
+
+String readNpmPackageScopeName() {
+	Object packageJson = readJSON file: 'package.json'
+	String[] nameParts = packageJson.name.tokenize('/')
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/configureArtifactRepository.groovy:23]: why do you need to tokenize here? aren't names like `symbol-sdk`?

**Review by Jaguar0625 — APPROVED** [review]

File: jenkins/shared-library/vars/configureArtifactRepository.groovy (hunk 3 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/configureArtifactRepository.groovy:23]: no, their packages have a scope(triton), e.g.  ``@triton/venues``

**Review by Jaguar0625 — APPROVED** [review]
👍

**Pull request merged**